spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/launcher/ModuleLauncher.java

```diff
@@ -92,6 +121,73 @@ public void launch(List<ModuleLaunchRequest> moduleLaunchRequests) {
 		return result;
 	}
 
+	public void aggregateAndLaunchModules(List<ModuleLaunchRequest> moduleLaunchRequests, final String[] parentArgs) {
+		try {
+			List<String> mainClassNames = new ArrayList<>();
+			List<URL> jarURLs = new ArrayList<>();
+			List<String> seenArchives = new ArrayList<>();
+			final List<String[]> arguments = new ArrayList<>();
+			// aggregate jars from all modules and extract their main Classes
+			for (ModuleLaunchRequest moduleLaunchRequest : moduleLaunchRequests) {
+				Resource resource = resolveModule(moduleLaunchRequest.getModule());
+				JarFileArchive jarFileArchive = new JarFileArchive(resource.getFile());
+				jarURLs.add(jarFileArchive.getUrl());
+				for (Archive archive : jarFileArchive.getNestedArchives(ArchiveMatchingEntryFilter.FILTER)) {
+					// avoid duplication based on unique JAR names
+					// TODO - read the metadata from the JARs, do proper version resolution on merge
+					String urlAsString = archive.getUrl().toString();
+					String jarNameWithExtension = urlAsString.substring(0, urlAsString.lastIndexOf("!/"));
+					String jarNameWithoutExtension = jarNameWithExtension.substring(jarNameWithExtension.lastIndexOf("/") + 1);
+					if (!seenArchives.contains(jarNameWithoutExtension)) {
+						seenArchives.add(jarNameWithoutExtension);
+						jarURLs.add(archive.getUrl());
+					}
+				}
+				mainClassNames.add(jarFileArchive.getMainClass());
+				arguments.add(toArgArray(moduleLaunchRequest.getArguments()));
+			}
+			final ClassLoader classLoader = new LaunchedURLClassLoader(jarURLs.toArray(new URL[jarURLs.size()]),
+					ClassloaderUtils.getExtensionClassloader());
+			final List<Class<?>> mainClasses = new ArrayList<>();
+			for (String mainClass : mainClassNames) {
+				mainClasses.add(ClassUtils.forName(mainClass, classLoader));
+			}
+			Runnable moduleAggregatorRunner = new Runnable() {
+				@Override
+				public void run() {
+					try {
+						// we expect the class and method to be found on the module classpath
+						Class<?> moduleAggregatorClass = ClassUtils.forName(MODULE_AGGREGATOR_CLASS, classLoader);
+						Method aggregateMethod = ReflectionUtils.findMethod(moduleAggregatorClass,
+								MODULE_AGGREGATOR_METHOD, String[].class, Class[].class, String[][].class);
+						aggregateMethod.invoke(null,
+								parentArgs,
+								mainClasses.toArray(new Class<?>[mainClasses.size()]) ,
+								arguments.toArray(new String[][]{}));
+					} catch (Exception e) {
+						log.error("Cannot start module group ", e);
```

Comment:
would change error message to "failed to start aggregated modules" (rather than use the word "group" here)